pkg/crypto/certloader/config.go

```diff
@@ -0,0 +1,209 @@
+	if certFile != "" && privkeyFile == "" {
+		return nil, ErrCertWithoutPrivkey
+	}
+	if certFile == "" && privkeyFile != "" {
+		return nil, ErrPrivkeyWithoutCert
+	}
+
+	cfg := &WatchedConfig{
+		log:         log,
+		caFiles:     caFiles,
+		certFile:    certFile,
+		privkeyFile: privkeyFile,
+	}
+
+	// load the files for the first time before starting the watcher and update
+	// goroutine.
+	if err := cfg.Reload(); err != nil {
+		return nil, err
+	}
+	if err := cfg.Watch(); err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// Watch initialize the WatchedConfig files watcher and update goroutine.
+func (cfg *WatchedConfig) Watch() error {
+	// FIXME: use sync.Once in order to run exactly one watcher.
+	// file watcher setup.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("Failed to create fsnotify watcher: %s", err)
+	}
+	for _, path := range cfg.caFiles {
+		if err := watcher.Add(path); err != nil {
+			watcher.Close()
+			return fmt.Errorf("Failed to add %q to fsnotify watcher: %s", path, err)
+		}
+	}
+	if cfg.certFile != "" {
+		if err := watcher.Add(cfg.certFile); err != nil {
+			watcher.Close()
+			return fmt.Errorf("Failed to add %q to fsnotify watcher: %s", cfg.certFile, err)
+		}
+	}
+	if cfg.privkeyFile != "" {
+		if err := watcher.Add(cfg.privkeyFile); err != nil {
+			watcher.Close()
+			return fmt.Errorf("Failed to add %q to fsnotify watcher: %s", cfg.privkeyFile, err)
+		}
+	}
+
+	cfg.stop = make(chan struct{})
+	go func() {
+		defer watcher.Close()
+		for {
+			select {
+			case event := <-watcher.Events:
+				log := cfg.log.WithFields(logrus.Fields{
+					logfields.Path: event.Name,
+					"operation":    event.Op,
+				})
+				log.Debug("Received fsnotify event")
+				switch event.Op {
+				case fsnotify.Create, fsnotify.Write, fsnotify.Chmod, fsnotify.Remove, fsnotify.Rename:
+					if err := cfg.Reload(); err != nil {
+						log.WithError(err).Warn("config update failed")
+					} else {
+						log.Info("config updated")
+					}
+				default:
+					log.Warn("unknown fsnotify event, ignoring.")
+				}
+			case err := <-watcher.Errors:
+				cfg.log.WithError(err).Warn("Watcher received an error")
+			case <-cfg.stop:
+				cfg.log.Info("Stopping fsnotify watcher")
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// CertificateAuthorityConfigured returns true when the WatchedConfig has at
+// least one CA file, false otherwise.
+func (cfg *WatchedConfig) CertificateAuthorityConfigured() bool {
+	return len(cfg.caFiles) > 0
+}
+
+// KeypairConfigured returns true when the WatchedConfig contains both a
+// certificate and its private key, false otherwise.
+func (cfg *WatchedConfig) KeypairConfigured() bool {
+	return cfg.certFile != "" && cfg.privkeyFile != ""
+}
+
+// Reload read the WatchedConfig files.
+func (cfg *WatchedConfig) Reload() error {
+	cfg.mu.Lock()
+	defer cfg.mu.Unlock()
+
+	// CA update
+	if cfg.CertificateAuthorityConfigured() {
+		if len(cfg.caFiles) > 0 {
```

Comment:
This `if` condition is essentially what `CertificateAuthorityConfigured()` does so you are actually checking for the same thing twice in a row, right? This also means that the `else` clause is unreachable.